Write comprehensive changelog for code in ICU4X so far; add more changelog docs (#7430)
unicode-org/icu4x: CHANGELOG.md

        - Improve Hijri docs (unicode-org#7330, unicode-org#7332, unicode-org#7333)
        - Remove `Hijri<AstronomicalSimulation>` simulation code, retaining hardcoded data, falling back to Tabular for non-modern dates (unicode-org#7301)
        - Replace `Date::day_of_week` by `Date::weekday` (unicode-org#7288)
        - Deprecate `Date::new_from_iso`/`Date::to_iso` (unicode-org#7287)
        - Optimize Hebrew and Julian calendars (unicode-org#7213)
        - Optimize day/week diffing to use RDs (unicode-org#7308)
    - `icu_casemap`
        - General changes only
    - `icu_collections`
        - Add `CodePointInversionListAndStringList::contains_utf8` (unicode-org#7363)
    - `icu_codepointtrie_builder`: `0.5.1 -> ???`
        - Remove serde dep from `icu_codepointtrie_builder` (unicode-org#7298)
    - `icu_collator`
        - General changes only
    - `icu_datetime`
        - Fix error handling for FieldSetBuilder (unicode-org#7245)
        - Add some support for `U` (cyclic year) datetime symbol (unicode-org#7328)
        - Remove old datetime data structs (unicode-org#7205)
        - Assorted improvements to icu_datetime docs (unicode-org#7244)
    - `icu_datetime`

>         - General changes only

Reviewer: I assume you mean:

```suggestion
    - `icu_decimal`
        - General changes only
```

    - `icu_experimental`: `0.4.0 -> ???`
        - `compactdecimal`
            - Don't hallucinate patterns (unicode-org#7387)
            - Pack compact decimal data (unicode-org#7388)
            - Round fractional compact decimals to two significant digits (unicode-org#7389)
        - `currency`
            - Add currency fractions provider (unicode-org#7278)
            - Update comments for currency data structures to enhance clarity (unicode-org#7405)
        - `dimension`
        - `measure`
        - `relativetime`
        - `units`
            - Correct region extraction for categorized display names (#7421)
    - `icu`
        - Add example for measuring Date::try_from_fields code size (unicode-org#7297)
    - `icu_list`
        - General changes only
    - `icu_locale`
        - Add docs discouraging direct conversion from Locale to DataLocale for locale fallback (unicode-org#7348)
    - `icu_locale_core`
        - Fix regional override `-u-rg` (unicode-org#7337) and regional subdivision `-u-sd` (unicode-org#7341) to fix region-priority data loading in other components

>         - (Breaking internal API) The `struct_keyword!` macro was changed to operate on references for conversions (unicode-orgunicode-org#7361)

Author: It's the type of thing where I want this in the changelog because _in case_ something breaks we should help people, but also it's not actually as big a deal as it sounds when we write "breaking" in the changelog.

# Changelog

## Unreleased

Fully filled in up to 30c187f4b7

- Components
    - General
        - Use HTTPS links in docs (unicode-org#7212)
        - Various internal cleanups
    - `icu_calendar`
        - Introduce a new, type-safe `Month` type, preferred over using month codes (unicode-org#7147)
        - Restrict the range of valid dates constructed via certain constructors (unicode-org#7219, unicode-org#7227)
        - Fix out-of-range bug during date arithmetic (unicode-org#7257)
        - Implement ISO 8601 date duration parsing (#7355)
        - Correctly produce `ethioaa` calendars from `CalendarAlgorithm` (unicode-org#7321)
        - Respect `-u-rg` in calendar resolution (unicode-org#7376)

>         - Pack `ArithmeticDate` (unicode-org#7220)

Reviewer 2: very internal

Author: Optimizes stack size

        - `AnyCalendar` cleanups and docs fixes (unicode-org#7223, unicode-org#7225)
        - Improve Hijri docs (unicode-org#7330, unicode-org#7332, unicode-org#7333)
        - Remove `Hijri<AstronomicalSimulation>` simulation code, retaining hardcoded data, falling back to Tabular for non-modern dates (unicode-org#7301)
        - Replace `Date::day_of_week` by `Date::weekday` (unicode-org#7288)
        - Deprecate `Date::new_from_iso`/`Date::to_iso` (unicode-org#7287)
        - Optimize Hebrew and Julian calendars (unicode-org#7213)
        - Optimize day/week diffing to use RDs (unicode-org#7308)

>         - Untangle Japanese calendars (unicode-org#7334)

Reviewer 2: also internal

    - `icu_casemap`
        - General changes only
    - `icu_collections`
        - Add `CodePointInversionListAndStringList::contains_utf8` (unicode-org#7363)
    - `icu_codepointtrie_builder`: `0.5.1 -> ???`
        - Remove serde dep from `icu_codepointtrie_builder` (unicode-org#7298)
    - `icu_collator`
        - General changes only
    - `icu_datetime`
        - Fix error handling for FieldSetBuilder (unicode-org#7245)
        - Add some support for `U` (cyclic year) datetime symbol (unicode-org#7328)
        - Remove old datetime data structs (unicode-org#7205)
        - Assorted improvements to icu_datetime docs (unicode-org#7244)
    - `icu_datetime`
        - General changes only
    - `icu_experimental`: `0.4.0 -> ???`
        - `compactdecimal`
            - Don't hallucinate patterns (unicode-org#7387)
            - Pack compact decimal data (unicode-org#7388)
            - Round fractional compact decimals to two significant digits (unicode-org#7389)

>             - Improve safety comment and test for PLURAL_PATTERN_0 (unicode-org#7391)

Reviewer 2: very internal. this PR only fixes up another PR that's also in this changelog, it's not a change that a client cares about

        - `currency`
            - Add currency fractions provider (unicode-org#7278)
            - Update comments for currency data structures to enhance clarity (unicode-org#7405)
        - `dimension`
        - `measure`
        - `relativetime`
        - `units`
            - Correct region extraction for categorized display names (#7421)
    - `icu`
        - Add example for measuring Date::try_from_fields code size (unicode-org#7297)
    - `icu_list`
        - General changes only
    - `icu_locale`
        - Fix `-u-sd` (unicode-org#7341)
        - Add docs discouraging direct conversion from Locale to DataLocale for locale fallback (unicode-org#7348)
    - `icu_locale_core`
        - Fix regional override `-u-rg` (unicode-org#7337) and regional subdivision `-u-sd` (unicode-org#7341) to fix region-priority data loading in other components
        - (Breaking internal API) The `struct_keyword!` macro was changed to operate on references for conversions (unicode-orgunicode-org#7361)
        - Add `From<&Value>` for struct preferences (unicode-org#7361)

>         - Correctly parse ethiopic-amete-alem alias (#7413)

Reviewer 2: ```suggestion
        - Correctly parse `-u-ca-ethiopic-amete-alem` alias (#7413)
```

        - Add `LocalePreferences::from_locale_strict` (unicode-org#7377)
        - Fix `-u-sd` (unicode-org#7341)
        - Fix `LanguageIdentifier::normalize_utf8` example (unicode-org#7372)
        - Use better types in `LocalePreferences` (unicode-org#7360)
    - `icu_normalizer`
        - Move `harfbuzz-traits` implementations into component crates (unicode-org#7200)
    - `icu_pattern`
        - Create SinglePlaceholderPattern::PASS_THROUGH (unicode-org#7393)
    - `icu_plurals`
        - FourBitMetadata should be checked against 0x10 or 0x0F, not 0x80 (unicode-org#7395)
        - Fix overflow in `PluralOperands` constructor (#7425)
    - `icu_properties`
        - Add enumerated property `Numeric_Type` (unicode-org#7157)
        - Add enumerated property `Joining_Group` (unicode-org#7293)
        - Add missing convenience API for `Basic_Emoji` and `EmojiSet` (unicode-org#7358)

>         - Stabilise `IndicConjuctBreak` (unicode-org#7280)

Reviewer 2: I think this is released

        - `AnyCalendar` cleanups and docs fixes (unicode-org#7223, unicode-org#7225)
        - Improve Hijri docs (unicode-org#7330, unicode-org#7332, unicode-org#7333)
        - Remove `Hijri<AstronomicalSimulation>` logic, retaining hardcoded data, improve docs (unicode-org#7301)
        - Replace `Date::day_of_week` by `Date::weekday` (unicode-org#7288)
        - Deprecate `Date::new_from_iso`/`Date::to_iso` (unicode-org#7287)
        - Optimize Hebrew and Julian calendars (unicode-org#7213)
        - Optimize day/week diffing to use RDs (unicode-org#7308)
        - Untangle Japanese calendars (unicode-org#7334)
    - `icu_casemap`
        - General changes only
    - `icu_collections`
        - Add `CodePointInversionListAndStringList::contains_utf8` (unicode-org#7363)
    - `icu_codepointtrie_builder`: `0.5.1 -> ???`
        - Remove serde dep from `icu_codepointtrie_builder` (unicode-org#7298)
    - `icu_collator`
        - General changes only
    - `icu_datetime`
        - Fix error handling for FieldSetBuilder (unicode-org#7245)
        - Add full support for `U` (cyclic year) datetime symbol (unicode-org#7328)
        - Remove old datetime data structs (unicode-org#7205)

>         - Clean up `FormattableAnyCalendar` (unicode-org#7318)

Reviewer: This is internal. Users will go to the docs not knowing what that is

            - Pack compact decimal data (unicode-org#7388)
            - Round fractional compact decimals to two significant digits (unicode-org#7389)
            - Improve safety comment and test for PLURAL_PATTERN_0 (unicode-org#7391)
        - `currency`
            - Add currency fractions provider (unicode-org#7278)
            - Update comments for currency data structures to enhance clarity (unicode-org#7405)
        - `dimension`
        - `measure`
        - `relativetime`
        - `units`
            - Correct region extraction for categorized display names (#7421)
    - `icu`
        - Add example for measuring Date::try_from_fields code size (unicode-org#7297)
    - `icu_list`
        - General changes only
    - `icu_locale`
        - Fix `-u-sd` (unicode-org#7341)
        - Resolve TODO around region fallback by adding docs about conversion between locale types (unicode-org#7348)
    - `icu_locale_core`
        - The `struct_keyword!` macro was changed to operate on references for conversions (unicode-orgunicode-org#7361)
        - Correctly parse ethiopic-amete-alem alias (#7413)
        - Add `LocalePreferences::from_locale_strict` (unicode-org#7377)

>         - Fix locale region override (unicode-org#7337)

Reviewer: This is the headline change for icu::locale. It was broken, as in not working at all, and it is now fixed, as in working as it is supposed to work.

>         - Fix locale region override (unicode-org#7337)

Reviewer: This is the headline change for icu::locale. It was broken, as in not working at all, and it is now fixed, as in working as it is supposed to work.

```suggestion
        - Fix regional override `-u-rg` (unicode-org#7337) and regional subdivision `-u-sd` (unicode-org#7341) to fix region-priority data loading in other components
```

and please move it to the top of the section.

>         - Fix `-u-sd` (unicode-org#7341)

Reviewer: Duplicate, and also I merged this in with the `-u-rg` one.

Reviewer 2: still there

>         - Add `From<&Value>` for struct preferences (unicode-org#7361)

Reviewer: I think we should highlight that this is a breaking (as in non-semver-compatible) change for anyone who might have been using the publicly exported macro `struct_keyword!`

EDIT: There actually is already a bullet point for this. Probably make it the second bullet point, below the regional override fix, and merge it with this one down here.

        - Stabilise `IndicConjuctBreak` (unicode-org#7280)
        - Constify `PropertyNamesLong`/`PropertNamesShort`/`PropertyParser` constructors (unicode-orgunicode-org#7294)
        - Fix script values (unicode-org#7269)
        - Move `harfbuzz-traits` implementations into component crates (unicode-org#7200)
        - Add conversions for `unicode_bidi::BidiClass` (unicode-org#7272)
        - Add conversions for properties/locale scripts (unicode-org#7270)
        - Validate properties names, constants (unicode-org#7284, unicode-org#7281)
    - `icu_segmenter`
        - Add non-complex segmenter constructors (unicode-org#7268)
        - (Experimental) Initial code for adaboost cjk word segmenter (unicode-org#7122, unicode-org#7217, unicode-org#7246, unicode-org#7344)
    - `icu_time`
        - Add docs for `DateTime`/`ZonedDateTime` semantics (unicode-org#7275)
        - Relax some bounds (unicode-org#7286)
- Data model and providers
    - `icu_provider_adapters`
        - General changes only
    - `icu_provider_baked`
        - Use `Index32` in baked data by default (unicode-org#7310)
    - `icu_provider_blob`
        - General changes only
    - `icu_provider`

>         - Breaking: Export an Encodeable instead of a VarULE and use Index32 (unicode-org#7310)

Reviewer: ```suggestion
        - Breaking: Add associated type to `MaybeEncodeAsVarULE` trait and use Index32 (unicode-org#7310)
```
